Add preemptible option on gce.py

curious-eyes opened the pull request:
Since libcloud 0.20.0 is released, we can use the ex_preemptible keyword. #2561

curious-eyes committed:
Add preemptible option on gce.py

robynbergeron commented:
Thanks @curious-eyes. @erjohnso please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

erjohnso commented:
Thanks for this @curious-eyes and sorry for the laggy review.

This looks good, but my only concern is that if a user is using libcloud < 0.20.0, that adding the extra param to the create_node method will cause a stacktrace.

I'd suggest adding a check if the new preemptible option is set, that you check the libcloud version and throw some kind of unsupported error and prompt the user to upgrade libcloud.  See [here](https://github.com/ansible/ansible-modules-core/blob/devel/cloud/google/gce.py#L338) for an example of checking the libcloud version.

curious-eyes commented:
Thanks for your fine advice, @erjohnso .

Added a check on the Libcloud version.
Besides, added dict for the optional arguments to the create_node method.

It will work well. Thanks again for reviewing.

erjohnso commented:
Thanks for addressing the feedback @curious-eyes.  The new changes look great and I think it passes Ansible guidelines.

shipit

@robynbergeron - this lgtm, please mark this for inclusion / merge.

gregdek commented:
Thanks again to @curious-eyes for this PR, and thanks @erjohnso for reviewing. Marking for inclusion.

bcoca commented on cloud/google/gce.py:
```diff
@@ -138,6 +138,11 @@
       - if set boot disk will be removed after instance destruction
     required: false
     default: "true"
+  preemptible:
```
needs version_added: "2.1"

bcoca commented:
you should make a note in the description that it needs libcloud 0.20 and above, as module requires only 0.13 (0.17 for some features)

robynbergeron commented:
Thanks @curious-eyes. @erjohnso please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

robynbergeron commented:
@curious-eyes -- thanks for the updates. @bcoca / @erjohnso -- let us know if it's ready to be shipped with the shipit comment, or a needs_revision comment if it needs further revision.

erjohnso commented on cloud/google/gce.py:
```diff
@@ -145,6 +145,13 @@
+    version_added: "2.1"
+    description:
+      - if set to true, instances will be preemptible and time-limited.
+        (requires libcloud >= 0.20.0)
```
@curious-eyes - I think Brian was also suggesting you also update the `requirements` section of the module docstring to suggest >= 0.20.0 is preferred for all features.

I think that's the final tweak and then you're all set.

curious-eyes commented:
Oh, I see! thanks @erjohnso @bcoca .

curious-eyes commented:
@bborysenko , thanks for your report. 

It seems to me that this is the same as #2761 .

gregdek commented:
Thanks @curious-eyes for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

curious-eyes commented:
@gregdek  I rebased PR.
ready_for_review

gregdek commented:
Thanks @curious-eyes. @erjohnso please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

gregdek commented:
Thanks @curious-eyes. @erjohnso please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

erjohnso commented:
Ok, @gregdek, this is ready and can be merged asap.  Thanks again for the fix @curious-eyes!

shipit

gregdek commented:
Thanks again to @curious-eyes for this PR, and thanks @erjohnso for reviewing. Marking for inclusion.

bcoca commented:
please cleanup your commit history, you have merges

gregdek commented:
Thanks @curious-eyes. To the current maintainers, @erjohnso please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

erjohnso commented:
Sorry @curious-eyes - looks like they'd like you to squash your commits and rebase so they can merge a single commit.

@gregdek - already reviewed and once @curious-eyes cleans it up, please merge it.

shipit

curious-eyes commented:
Squashed changes to single commit. Thanks @bcoca @gregdek @erjohnso.

gregdek commented:
Thanks again to @curious-eyes for this PR, and thanks @erjohnso for reviewing. Marking for inclusion.

erjohnso commented:
Thanks @gregdek!  And thanks again for the fix @curious-eyes!
